File: users/src/ports/users_repository.rs
use uuid::Uuid;

use super::super::entities::user::User;

pub trait UsersRepository: Send + Sync {
    fn create(&self, user: User) -> Result<(), String>;
    fn read(&self, id: Uuid) -> Result<Option<User>, String>;
    fn read_by_email(&self, email: String) -> Result<Option<User>, String>;
    fn update(&self, user: User) -> Result<(), String>;
    fn delete(&self, id: Uuid) -> Result<(), String>;
    fn clone_box(&self) -> Box<dyn UsersRepository + Send + Sync>;
}

impl Clone for Box<dyn UsersRepository + Send + Sync> {
    fn clone(&self) -> Box<dyn UsersRepository + Send + Sync> {
        self.clone_box()
    }
}

File: users/src/ports/users.repository.ts
import { Email } from 'src/entities/emails';
import { Identity } from 'src/entities/identities';
import { User } from 'src/entities/users';

export interface UsersRepository {
  create(user: User): Promise<User>;
  readByIdentity(identity: Identity): Promise<User | undefined>;
  readByEmail(email: Email): Promise<User | undefined>;
  update(user: User): Promise<User>;
  delete(user: User): Promise<User>;
}

export class SaveUserError extends Error {}

export class ReadUserError extends Error {}

export class DeleteUserError extends Error {}
